File: src/main/java/com/tommykvant/wikivoyage/parsers/ContentParser.java
package com.tommykvant.wikivoyage.parsers;

import com.tommykvant.wikivoyage.creators.ListCreator;
import com.tommykvant.wikivoyage.details.content.Content;
import com.tommykvant.wikivoyage.details.content.HorizontalSpace;
import com.tommykvant.wikivoyage.details.data.IndentedTextContent;
import com.tommykvant.wikivoyage.details.data.TextContent;

import java.util.ArrayList;

public class ContentParser {

    public static ArrayList<Content> parse(StringIterator iterator) {
        ArrayList<Content> out = new ArrayList<Content>();

        while (iterator.hasNext() && iterator.peekNext() != '=') {
            if (iterator.peekNext() == '\n') {
                // deal with horizontal spacing
                out.add(new HorizontalSpace());
                iterator.next();
            } else if (iterator.peekNext() == '*') {
                // Parse lists separately
                out.add(ListCreator.create(iterator, out));
            } else if (iterator.peekNext() == ':') {
                // Indentation
                out.add(new IndentedTextContent(iterator, out));
            } else {
                // String text = parseText(iterator);
                out.add(new TextContent(iterator, out));
            }
        }

        return out;
    }
}

File: WikiVoyageParser/src/main/kotlin/com.arclights.wikivoyageparser/parser/ContentParser.kt
package com.arclights.wikivoyageparser.parser

import com.arclights.wikivoyageparser.Constants.Companion.headerDelimiter
import com.arclights.wikivoyageparser.Content
import com.arclights.wikivoyageparser.HorizontalSpace
import com.arclights.wikivoyageparser.StringIterator
import java.util.*

object ContentParser {
    fun parse(iterator: StringIterator): List<Content> {
        val out = ArrayList<Content>()

        while (iterator.hasNext() && iterator.peekNext() != headerDelimiter) {
            when (iterator.peekNext()) {
                '\n' -> {
                    // deal with horizontal spacing
                    out.add(HorizontalSpace())
                    iterator.next()
                }
                '*' -> out.add(ListParser.parse(iterator))
                ':' -> println("Found indented content: ${iterator.currentPlace()}")
                '{' -> out.add(TemplateParser.parse(iterator))
                else -> out.add(TextParser.parseText(iterator))
            }
        }

        return out
    }
}